### PR TITLE
http: use typed headers in encoding path

### PR DESCRIPTION
**Pull request by mattklein123**
This is "the big one" in the header map refactor. This commit uses
typed headers all the way through the encoding path and thus requires
an interface change in HTTP filters.

Although the lines of code changed in this PR is huge, the regression
risk is low as this change is almost entirely an extremely tedious fixup
PR to use the correct types. Quite a few filter tests needed fixing as
they were using the wrong header type for trailers, etc.

Risk Level: Low
Testing: Existing tests
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
@lizan as maintainer on-call I'm going to assign this to you, sorry. :) It's mostly an extremely mechanical and boring PR, but I would still appreciate a full review to make sure I didn't miss anything obvious. Thank you!

**Comment by mattklein123** [comment]
This is failing due to a needed change in the filter example repo. @lizan how are we handling that these days?

**Comment by lizan** [comment]
We merge a PR in filter example first, and then update the filter example SHA in CI script.

File: source/common/grpc/async_client_impl.cc [hunk 1]
```diff
@@ -97,18 +97,18 @@ void AsyncStreamImpl::initialize(bool buffer_body_for_retry) {
...
   if (http_response_status != enumToInt(Http::Code::OK)) {
     // https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md requires that
     // grpc-status be used if available.
     if (end_stream && grpc_status) {
-      // There is actually no use-after-move problem here,
-      // because it will only be executed when end_stream is equal to true.
-      onTrailers(std::move(headers)); // NOLINT(bugprone-use-after-move)
+      // Due to headers/trailers type differences we need to copy here. This is an uncommon case but
+      // we can potentially optimize in the future.
+      onTrailers(Http::createHeaderMap<Http::ResponseTrailerMapImpl>(*headers));
```
Comment (lizan) [inline source/common/grpc/async_client_impl.cc]: can we extract the relevant part of `onTrailers` to `onGrpcTrailers(HeaderMapImplPtr)` and call them from both here and `onTrailers`?

Comment (mattklein123) [reply source/common/grpc/async_client_impl.cc]: The issue here is that we have to change the physical type (turn headers into trailers) and I don't think extraction will help here? I was able to split out the logic in the transcoder to deal with a similar case. With all of that said, the next step is to move towards HeaderMap being an abstract base class, with relevant O(1) headers moved to where they are used, so I'm going to need to come back to this soon (gRPC headers/trailers confuse this a bit).

**Comment by mattklein123** [comment]
@lizan this should be good to go. LMK if you have any other concerns. I would love to get this in before I lose another merge race.

File: source/common/grpc/async_client_impl.cc (hunk 1 above)
Comment (lizan) [reply source/common/grpc/async_client_impl.cc]: OK sounds fair.

**Pull request merged**